### PR TITLE
Fix #6359

### LINKED ISSUE #6359: python: /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:106: static bool llvm::isa_impl_cl<To, const From*>::doit(const From*) [with To = clang::UsingDecl; From = clang::Decl]: Assertion `Val && "isa<> used on a null pointer"' failed.
Hi, I am running nightly tests of COOL which is built using PyROOT, and since the 10th September, one test is failing when using ROOT master branch on all the platforms we test (slc6 with gcc8 and centos7 with gcc8-10 and clang10).
I suspect one commit from Sep 9, 2020 to be responsible for this change of behaviour.

Here is the error and backtrace:
```
      python: /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:106: static bool llvm::isa_impl_cl<To, const From*>::doit(const From*) [with To = clang::UsingDecl; From = clang::Decl]: Assertion `Val && "isa<> used on a null pointer"' failed.
       *** Break *** abort
      ===========================================================
      There was a crash (#7 0x00007facdfe876c8 in SigHandler(ESignals) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/unix/src/TUnixSystem.cxx:407).
      This is the entire stack trace of all threads:
      ===========================================================
      #0  0x00007facdeaa889e in waitpid () from /lib64/libc.so.6
      #1  0x00007facdea3a4e9 in do_system () from /lib64/libc.so.6
      #2  0x00007facdfe8b434 in TUnixSystem::Exec(char const*) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/unix/src/TUnixSystem.cxx:2120
      #3  0x00007facdfe8bc6a in TUnixSystem::StackTrace() () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/unix/src/TUnixSystem.cxx:2411
      #4  0x00007facdc599248 in (anonymous namespace)::do_trace(int) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/bindings/pyroot/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx:182
      #5  0x00007facdc5992c1 in (anonymous namespace)::TExceptionHandlerImp::HandleException(int) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/bindings/pyroot/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx:195
      #6  0x00007facdfe8f519 in TUnixSystem::DispatchSignals(ESignals) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/unix/src/TUnixSystem.cxx:3644
      #7  0x00007facdfe876c8 in SigHandler(ESignals) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/unix/src/TUnixSystem.cxx:407
      #8  0x00007facdfe8f474 in sighandler(int) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/unix/src/TUnixSystem.cxx:3620
      #9  <signal handler called>
      #10 0x00007facdea2e4f5 in raise () from /lib64/libc.so.6
      #11 0x00007facdea2fcd5 in abort () from /lib64/libc.so.6
      #12 0x00007facdea2766e in __assert_fail_base () from /lib64/libc.so.6
      #13 0x00007facdea27730 in __assert_fail () from /lib64/libc.so.6
      #14 0x00007facd7a7438f in llvm::isa_impl_cl<clang::UsingDecl, clang::Decl const*>::doit(clang::Decl const*) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:106
      #15 0x00007facd7a74211 in llvm::isa_impl_wrap<clang::UsingDecl, clang::Decl const*, clang::Decl const*>::doit(clang::Decl const* const&) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:133
      #16 0x00007facd7a7407a in llvm::isa_impl_wrap<clang::UsingDecl, clang::Decl const* const, clang::Decl const*>::doit(clang::Decl const* const&) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:124
      #17 0x00007facd7a73d65 in bool llvm::isa<clang::UsingDecl, clang::Decl const*>(clang::Decl const* const&) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:144
      #18 0x00007facd7a73b0c in llvm::cast_retty<clang::UsingDecl, clang::Decl const*>::ret_type llvm::dyn_cast<clang::UsingDecl, clang::Decl const>(clang::Decl const*) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/interpreter/llvm/src/include/llvm/Support/Casting.h:334
      #19 0x00007facd7a73536 in TClingMemberIter::Advance() () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/metacling/src/TClingMemberIter.cxx:128
      #20 0x00007facd7a780e9 in TClingMemberIter::Init() () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/metacling/src/TClingMemberIter.h:152
      #21 0x00007facd7a763a5 in TClingMethodInfo::TClingMethodInfo(cling::Interpreter*, TClingClassInfo*) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/metacling/src/TClingMethodInfo.cxx:276
      #22 0x00007facd79333fe in TCling::MethodInfo_Factory(ClassInfo_t*) const () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/metacling/src/TCling.cxx:8863
      #23 0x00007facdfe45a3a in TListOfFunctions::Load() () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/meta/src/TListOfFunctions.cxx:390
      #24 0x00007facdfe18843 in TClass::GetListOfMethods(bool) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/core/meta/src/TClass.cxx:3721
      #25 0x00007facdc59f206 in Cppyy::GetNumMethods(unsigned long) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/bindings/pyroot/cppyy/cppyy-backend/clingwrapper/src/clingwrapper.cxx:1323
      #26 0x00007facdc8effb9 in CPyCppyy::BuildScopeProxyDict(unsigned long, _object*) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:166
      #27 0x00007facdc8f2654 in CPyCppyy::CreateScopeProxy () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/bindings/pyroot/cppyy/CPyCppyy/src/ProxyWrappers.cxx:673
      #28 0x00007facdc8cbbe2 in CPyCppyy::meta_getattro(_object*, _object*) () at /workspace/build/projects/ROOT-HEAD/src/ROOT/HEAD/bindings/pyroot/cppyy/CPyCppyy/src/CPPScope.cxx:323
      #29 0x00007facdf52e1cc in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #30 0x00007facdf5342fb in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #31 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #32 0x00007facdf4adf96 in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #33 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #34 0x00007facdf52bead in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #35 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #36 0x00007facdf4adedc in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #37 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #38 0x00007facdf49253c in instancemethod_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #39 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #40 0x00007facdf4e8a6a in slot_tp_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #41 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #42 0x00007facdf52fd57 in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #43 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #44 0x00007facdf4adf96 in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #45 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #46 0x00007facdf52bead in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #47 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #48 0x00007facdf4adedc in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #49 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #50 0x00007facdf49253c in instancemethod_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #51 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #52 0x00007facdf4e8a6a in slot_tp_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #53 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #54 0x00007facdf52fd57 in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #55 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #56 0x00007facdf4adf96 in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #57 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #58 0x00007facdf52bead in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #59 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #60 0x00007facdf4adedc in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #61 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #62 0x00007facdf49253c in instancemethod_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #63 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #64 0x00007facdf4e8a6a in slot_tp_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #65 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #66 0x00007facdf52fd57 in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #67 0x00007facdf5342fb in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #68 0x00007facdf5342fb in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #69 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #70 0x00007facdf4adf96 in function_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #71 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #72 0x00007facdf49253c in instancemethod_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #73 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #74 0x00007facdf4e8fd2 in slot_tp_init () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #75 0x00007facdf4e51fa in type_call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #76 0x00007facdf4845e3 in PyObject_Call () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #77 0x00007facdf52fd57 in PyEval_EvalFrameEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #78 0x00007facdf534aa2 in PyEval_EvalCodeEx () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #79 0x00007facdf534d39 in PyEval_EvalCode () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #80 0x00007facdf5597a8 in PyRun_FileExFlags () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #81 0x00007facdf55a723 in PyRun_SimpleFileExFlags () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #82 0x00007facdf56e3da in Py_Main () from /cvmfs/sft-nightlies.cern.ch/lcg/latest/Python/2.7.16-3adfa/x86_64-slc6-gcc8-dbg/bin/../lib/libpython2.7.so.1.0
      #83 0x00007facdea1ad20 in __libc_start_main () from /lib64/libc.so.6
      #84 0x00000000004006a1 in _start ()
```

Let me know if I can provide more related details,
Cheers,
Charles

### PR DESCRIPTION
**Pull request by Axel-Naumann**
fixes #6359

### CODE REVIEW

**Comment by Axel-Naumann** [comment]
@phsft-bot build

File: core/metacling/src/TClingMethodInfo.cxx [hunk 1]
```diff
@@ -265,8 +265,10 @@ TClingMethodInfo::TClingMethodInfo(cling::Interpreter *interp,
 
       // Assemble special functions (or FunctionTemplate-s) that are synthesized from DefinitionData but
       // won't be enumerated as part of decls_begin()/decls_end().
-      auto Ctors = SemaRef.LookupConstructors(CXXRD);
-      SpecFuncs.append(Ctors.begin(), Ctors.end());
+      for (clang::NamedDecl *ctor : SemaRef.LookupConstructors(CXXRD)) {
+         if (auto *CXXMD = llvm::dyn_cast<clang::CXXMethodDecl>(ctor))
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.cxx:269]: Could you add a comment expressing what this is intended to filter out?

File: core/metacling/src/TClingMethodInfo.cxx [hunk 2]
```diff
@@ -265,8 +265,11 @@ TClingMethodInfo::TClingMethodInfo(cling::Interpreter *interp,
 
       // Assemble special functions (or FunctionTemplate-s) that are synthesized from DefinitionData but
       // won't be enumerated as part of decls_begin()/decls_end().
-      auto Ctors = SemaRef.LookupConstructors(CXXRD);
-      SpecFuncs.append(Ctors.begin(), Ctors.end());
+      for (clang::NamedDecl *ctor : SemaRef.LookupConstructors(CXXRD)) {
+         // Filter out constructor templates, they are not functions we can iterate over:
+         if (auto *CXXMD = llvm::dyn_cast<clang::CXXMethodDecl>(ctor))
```
Comment (pcanal) [inline core/metacling/src/TClingMethodInfo.cxx:270]:
```suggestion
         if (auto *CXXMD = llvm::dyn_cast<clang::CXXConstructorDecl>(ctor))
```

Comment (Axel-Naumann) [reply core/metacling/src/TClingMethodInfo.cxx]: > or maybe we also need CXXConversionDecl.

Conversion functions are not constructors (they create the return type, not their class type).

> for reference what is the Clang type of  "constructor templates"

`FunctionTemplateDecl`, unsurprisingly

**Review by pcanal — APPROVED** [review]

**Pull request merged**